### PR TITLE
executor: Let the SHUTDOWN statement do a graceful shutdown

### PR DESCRIPTION
**Pull request by dveeden**
### What problem does this PR solve?

When sending `SIGTERM` to `tidb-server` instead of killing it
this allows graceful shutdown to work.

Details on graceful shutdown:
https://github.com/pingcap/tidb/blob/749a99c02027fb43406d01ed9b0f33f219a37db2/config/config.toml.example#L64-L66

Problem Summary:

The behavior of stopping TiDB via the UNIX kill command (which sends `SIGTERM` by default) or by using the SHUTDOWN statement as different. Both should do a graceful shutdown.

### What is changed and how it works?

What's Changed:

This changes the signal send to `tidb-server` when the `SHUTDOWN` statement is triggered.

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

- Run `tidb-server` with `graceful-wait-before-shutdown = 30`
- Run a `SHUTDOWN` statement and inspect the logs.

### Release note 

- The `SHUTDOWN` statement now does a graceful shutdown

### CODE REVIEW

**Comment by dveeden** [comment]
cc @morgo

**Review by jackysp — COMMENTED** [review]
LGTM

**Comment by lysu** [comment]
LGTM

but it seems we miss check privilege for this command?

**Comment by dveeden** [comment]
> LGTM
> 
> but it seems we miss check privilege for this command?

<pre><span style="background-color:#E4E4E4"><font color="#005F5F">5.7.25-TiDB-v5.2.0-alpha-135-g6db6111e6</font></span><span style="background-color:#3A3A3A"><font color="#E4E4E4"></font></span><span style="background-color:#3A3A3A"><font color="#FDF6E3"> 127.0.0.1:4000 </font></span><span style="background-color:#6C6C6C"><font color="#3A3A3A"></font></span><span style="background-color:#6C6C6C"><font color="#FDF6E3"> test </font></span><span style="background-color:#D75F00"><font color="#6C6C6C"></font></span><span style="background-color:#D75F00"><font color="#FDF6E3"> SQL </font></span><span style="background-color:#073642"><font color="#D75F00"> </font></span>create user &apos;u&apos;@&apos;%&apos; identified by &apos;uuu&apos;;
Query OK, 0 rows affected (0.0050 sec)
</pre>

<pre>[dvaneeden@dve-carbon ~]$ mysqlsh --quiet-start=2 mysql://u:uuu@127.0.0.1:4000/
<span style="background-color:#E4E4E4"><font color="#005F5F">5.7.25-TiDB-v5.2.0-alpha-135-g6db6111e6</font></span><span style="background-color:#3A3A3A"><font color="#E4E4E4"></font></span><span style="background-color:#3A3A3A"><font color="#FDF6E3"> 127.0.0.1:4000 </font></span><span style="background-color:#D75F00"><font color="#3A3A3A"></font></span><span style="background-color:#D75F00"><font color="#FDF6E3"> SQL </font></span><span style="background-color:#073642"><font color="#D75F00"> </font></span>shutdown;
ERROR: 8121 (HY000): privilege check for &apos;SHUTDOWN&apos; fail
</pre>

To me it looks like the privilege check works fine, but please double check.

**Review by morgo — CHANGES_REQUESTED** [review]
`SIGTERM` can be ignored, and I am concerned there are scenarios where this will fail to work as expected like this one: https://github.com/pingcap/tidb/issues/18336

Can you modify it to attempt SIGTERM and upon timeout revert to the previous behavior?

**Comment by dveeden** [comment]
side note: https://github.com/golang/go/issues/9764 is preventing me from adding logging about what process is sending the `SIGTERM` signal. I don't think adding Linux specific code to do this is reasonable at this point.

File: executor/simple.go [hunk 1]
```diff
@@ -1556,7 +1557,22 @@ func (e *SimpleExec) executeShutdown(s *ast.ShutdownStmt) error {
...
+
+	// Sending SIGKILL should not be needed as SIGTERM should cause a graceful shutdown after
+	// n seconds as configured by the GracefulWaitBeforeShutdown. This is here in case that doesn't
+	// work for some reason.
+	graceTime := config.GetGlobalConfig().GracefulWaitBeforeShutdown
+	if graceTime < 5 {
+		graceTime = 5
+	}
+	time.Sleep(time.Second * time.Duration(graceTime*2))
+	logutil.BgLogger().Info("Killing process as grace peroid is over", zap.Int("pid", p.Pid))
```
Comment (morgo) [inline executor/simple.go:1574]:
```suggestion
	logutil.BgLogger().Info("Killing process as grace period is over", zap.Int("pid", p.Pid), zap.Init("graceTime", graceTime))
```

File: executor/simple.go [hunk 2]
```diff
@@ -1556,7 +1557,22 @@ func (e *SimpleExec) executeShutdown(s *ast.ShutdownStmt) error {
...
+	}
+
+	// Sending SIGKILL should not be needed as SIGTERM should cause a graceful shutdown after
+	// n seconds as configured by the GracefulWaitBeforeShutdown. This is here in case that doesn't
+	// work for some reason.
+	graceTime := config.GetGlobalConfig().GracefulWaitBeforeShutdown
+	if graceTime < 5 {
+		graceTime = 5
+	}
+	time.Sleep(time.Second * time.Duration(graceTime*2))
```
Comment (morgo) [inline executor/simple.go:1573]: Why wait double the grace period? I think it is more predictable to wait exactly the graceTime.

Comment (dveeden) [reply executor/simple.go:1573]: When this works as it should the real shutdown starts after the grace period. I don't want to kill the process while the shutdown is happening. So either we wait for `graceTime+30` or for `graceTime*2` or so.

**Review by morgo — APPROVED** [review]

**Comment by morgo** [comment]
Thanks for adjusting this PR. Just to summarize discussion for the issue:

- Real shutdown has to integrate with graceful shutdown, but there needs to be some delay between the two. 
- This could be a function of gracefulTime * N or gracefulTime + N. My preference was for the latter.

**Comment by dveeden** [comment]
PTAL @pingcap/exec-reviewers

**Review by lysu — APPROVED** [review]

**Review by lzmhhh123 — APPROVED** [review]

**Pull request merged**